pallets/sponsorship/src/lib.rs

```diff
@@ -0,0 +1,630 @@
+		/// This would allow a user to bypass the instance filter or alter the origin of the calls.
+		type SponsorshipType: Parameter
+			+ Member
+			+ Ord
+			+ PartialOrd
+			+ InstanceFilter<<Self as Config>::RuntimeCall>
+			+ MaxEncodedLen
+			+ Default;
+		/// Type representing the weight of this pallet
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::storage]
+	/// Details of a pot.
+	pub(super) type Pot<T: Config> = StorageMap<_, Blake2_128Concat, T::PotId, PotDetailsOf<T>, OptionQuery>;
+
+	#[pallet::storage]
+	/// User details of a pot.
+	pub(super) type User<T: Config> =
+		StorageDoubleMap<_, Blake2_128Concat, T::PotId, Blake2_128Concat, T::AccountId, UserDetailsOf<T>, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Event emitted when a new pot is created.
+		PotCreated(T::PotId),
+		/// Event emitted when a pot is removed.
+		PotRemoved(T::PotId),
+		/// Event emitted when user/users are registered indicating the list of them
+		UsersRegistered(T::PotId, Vec<T::AccountId>),
+		/// Event emitted when user/users are removed indicating the list of them
+		UsersRemoved(T::PotId, Vec<T::AccountId>),
+		/// Event emitted when a sponsor_me call has been successful indicating the reserved amount
+		Sponsored { top_up: BalanceOf<T>, refund: BalanceOf<T> },
+		/// Event emitted when the transaction fee is paid showing the payer and the amount
+		TransactionFeePaid(T::AccountId, BalanceOf<T>),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The pot ID is already taken.
+		InUse,
+		/// The signing account has no permission to do the operation.
+		NoPermission,
+		/// The pot does not exist.
+		PotNotExist,
+		/// The user is not registered for the pot.
+		UserNotRegistered,
+		/// The user is already registered for the pot.
+		UserAlreadyRegistered,
+		/// The user is not removable due to holding some reserve.
+		CannotRemoveProxy,
+		/// Logic error: cannot create proxy account for user.
+		/// This should never happen.
+		CannotCreateProxy,
+		/// Balance leak is not allowed during a sponsored call.
+		BalanceLeak,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a new sponsorship pot and set its limits.
+		/// The pot id shouldn't be in use.
+		///
+		/// Emits `PotCreated(pot)` event when successful.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::create_pot())]
+		pub fn create_pot(
+			origin: OriginFor<T>,
+			pot: T::PotId,
+			sponsorship_type: T::SponsorshipType,
+			fee_quota: BalanceOf<T>,
+			reserve_quota: BalanceOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(!Pot::<T>::contains_key(pot), Error::<T>::InUse);
+
+			<Pot<T>>::insert(
+				pot,
+				PotDetailsOf::<T> {
+					sponsor: who,
+					sponsorship_type,
+					remained_fee_quota: fee_quota,
+					remained_reserve_quota: reserve_quota,
+				},
+			);
+
+			Self::deposit_event(Event::PotCreated(pot));
+			Ok(())
+		}
```

Comment:
I think this extrinsic should reserve `fee_quota.saturating_add(reserve_quota)` to:
1. Avoid spamming of this extrinsic
2. Ensure the sponsor has enough funds available for the limits they are setting themselves